### DISCUSSION TITLE
Allow Access to MPI Function Pointers

### DISCUSSION THREAD

**Discussion by FreddieWitherden** [mpi4py/mpi4py#587] [Q&A] [answered]
It would be nice if mpi4py could expose function pointers to the underlying MPI functions it has wrapped.

Use case: our code employs JIT compilation + task graphs.  We would like to kick-off an MPI function (e.g. `MPI_Start`) after the graph has finished executing a node (kernel).  At the moment this requires a costly callback into Python.  However, if we could grab the function pointers then we could pass a pointer to `MPI_Start` down into our JIT'ed code along with the relevant MPI_Request objects (which we create and manage at the Python level and I *think* should be able to get the address of easily enough).

Historically we've not tried to do this due to ABI issues, but with MPI slowly converging on an ABI linkless MPI is starting to become viable where all our JIT code needs is the function pointers themselves (which mpi4py knows).

**Comment by dalcinl** [comment, marked as answer]
I agree, it would be great, but putting all this in mpi4py's shoulders may involve a lot of implementation work I cannot afford, not to mention future maintenance if the thing would be contributed by someone.

IMHO, all what you request can be easily done with the help of `ctypes` and a few private APIs that mpi4py already exposes. For example, to the the address of `MPI_Start` and the pointer to the `MPI_Request` handle within a `MPI.Request` Python object, you could simply do the following:

```python
import ctypes
from mpi4py import MPI

mod =  ctypes.CDLL(MPI.__file__)
MPI_Start = mod.MPI_Start
request = MPI.Request()

MPI_Start_addr = ctypes.cast(MPI_Start, ctypes.c_void_p).value
request_handle = request.handle  # unused, just to highlight its existence
request_addr =  MPI._addressof(request)
print(MPI_Start_addr, request_addr)
```


PS: To workaround the usual `int` vs `void*` of MPICH vs Open MPI, look at [`demo/wrap-ctypes`](https://github.com/mpi4py/mpi4py/tree/master/demo/wrap-ctypes) for the use of `MPI._sizeof()`

**Reply by FreddieWitherden** [reply]
Are the `__file__` and `_addressof` members stable?  We'd be more than happy to use ctypes to pull out the members we want although try to avoid using private/unstable members where possible.

**Reply by dalcinl** [reply]
1. `__file__` is a standard Python attribute for modules set by the import system [[docs]](https://docs.python.org/3/reference/datamodel.html#module.__file__). Alternatively, you can use `MPI.__spec__.origin`. Obviously, I cannot make any promises about the stability this member. However, it has been present at least since Python 2.3 (the first version I ever used).
2. The `MPI._addressof()` and `MPI._sizeof()` functions are just a little private functions specific to mpi4py. Being private is just a lame attempt of mine to keep uninformed users form messing with some delicate things like internal MPI handles. For example, you could abuse of the handkle address to modify `MPI.COMM_WORLD` and set it to something else. I'm not interested in having issue reports about such hacks, I already have enough dealing with bugs in MPI implementations.  Another reason to have it (conventionally, by name) private it simply that these functions are not standard MPI APIs, but a tool designed to do very low level things exactly like what you need to do. All that being said, these private functions are indeed stable. I don't break APIs just for fun, I will keep these functions around even if some day  find a better public way to do all this.
